debug: tracing: Add CPU stats module.

MarekPieta opened the pull request:
Change adds CPU stats module for tracing hooks (compares time spend in idle thread, non idle threads and scheduler). Feature uses hooks for threads switching in/out and interrupt enter/exit.

aescolar commented on subsys/debug/tracing/cpu_stats.c:
```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
```
Zephyr follows Apache 2 not BSD with extras.

MarekPieta replied:
fixed

aescolar commented on include/tracing.h:
```diff
@@ -26,8 +26,13 @@ void z_sys_trace_thread_switched_in(void);
 void z_sys_trace_thread_switched_out(void);
 #endif
 
+
```
extra \n

MarekPieta replied:
removed

aescolar reviewed (dismissed):
So far, I have just browsed fast thru the code, and saw that license thing.

aescolar commented:
It would be nice to have a bit more descriptive commit message

pizi-nordic reviewed (commented):
Could we have this functionality without using tracing API? It is designed to something else in my opinion.

We already have logger and console. With little effort we can either print CPU usage periodically or have an equivalent of "ps" and "top" commands in Zephyr, which seems to be much better solution.

mped-oticon commented on subsys/debug/tracing/cpu_stats.c:
```diff
@@ -0,0 +1,176 @@
+#endif
+}
+
+void increase_counter(volatile u64_t *cnt)
```
So accumulate the deltas (which is the same as just the linear time sampled), OK.
Another way is to just count the u32 wraps, then you need less arithmetic here (more efficient) and you can avoid 64bit arith. Then it is up to you if you want to move the 64 bit arith to the presentation printk.

MarekPieta replied:
Counting u32 wraps could make us lose accuracy, which is important for performance measurements. It is why i decided to use 64 bits.

mped-oticon reviewed (changes requested):
This code calculates since-boot average cpu usage, which may not be very interesting.
I'd say some windowing should be applied.
Other minor comments.

mped-oticon commented:
@pizi-nordic 

> Could we have this functionality without using tracing API? It is designed to something else in my opinion.

I agree, but I can understand why this PR is made as it; the tracing API is a convenient way to hook in.
Also this PR does not interfere with the existing traces - so it appears safe to do. I don't mind it :)

> We already have logger and console. With little effort we can either print CPU usage periodically or have an equivalent of "ps" and "top" commands in Zephyr, which seems to be much better solution.

Also agree. Though it requires a shell etc etc. Oticon will rather let a tool consume the tiny trace events and infer CPU usage from those - permitting to changing the windowing on post-fact.

MarekPieta commented on subsys/debug/tracing/cpu_stats.c:
```diff
@@ -0,0 +1,176 @@
+u32_t cpu_meas_sched_get(void)
+{
+	cpu_meas_update_counters();
+	return SYS_CLOCK_HW_CYCLES_TO_NS(sched_cnt) / 1000000;
```
Windowing is applied by resetting counters periodically or calling function from API (`cpu_meas_reset_counters`) - I will change commit message. 

Division here is to change ns to ms - so result will be more human readable (i will add comment to clarify this one). SYS_CLOCK_HW_CYCLES_TO_NS (x) is macro from Zephyr kernel API which gives us result time in ns - so I think it is portable. 
Mentioned divisions are performed rather rarely (once per providing result to user), so maybe it's not a problem.

`cpu_meas_sched_get` is part of API for user (same as `cpu_meas_idle_get` and `cpu_meas_non_idle_get`). I wanted to make user able to use code in two ways: delayed work printing percent of CPU usage periodically (with configurable period) or using API - for example to measure CPU usage in some important section for the code (user can reset counters, make measurement and then get results in ms for scheduler, idle thread, non idle threads). I think it may be convenient for users.

pdunaj replied:
@MarekPieta you could change the api to return times in ns (e.g. cpu_stats_sched_get_ns) and let the use do the division if needed. I know we use it on target but it may be that somebody would simply pass the values outside of the target.
Unless you would like to make sure we fit in u32_t... (maybe this should be extended too then)

MarekPieta commented on subsys/debug/tracing/cpu_stats.c:
```diff
@@ -0,0 +1,176 @@
+	return SYS_CLOCK_HW_CYCLES_TO_NS(sched_cnt) / 1000000;
+}
+
+u32_t cpu_meas_idle_get(void)
```
Part of API for users - mentioned above

MarekPieta commented on subsys/debug/tracing/cpu_stats.c:
```diff
@@ -0,0 +1,176 @@
+	return SYS_CLOCK_HW_CYCLES_TO_NS(idle_cnt) / 1000000;
+}
+
+u32_t cpu_meas_non_idle_get(void)
```
Part of API for users - mentioned above

MarekPieta commented on subsys/debug/Kconfig:
```diff
@@ -188,4 +188,21 @@ config OPENOCD_SUPPORT
+
+config TRACING_CPU_STATS_LOG
+	bool "Logging current CPU usage"
+	default n
```
ok, removed

MarekPieta commented on subsys/debug/Kconfig:
```diff
@@ -188,4 +188,21 @@ config OPENOCD_SUPPORT
 
+config TRACING_CPU_STATS
+	bool "Tracing CPU usage"
+	default n
```
ok, removed

MarekPieta committed:
debug: tracing: Add CPU stats module

Change adds CPU stats module for tracing hooks. Module provides
information about percent of CPU usage based on tracing hooks
for threads switching in and out, interrupts enters and exits.
cpu_stats only distinguishes between idle thread, non idle
thread and scheduler.

Signed-off-by: Marek Pieta <Marek.Pieta@nordicsemi.no>

dbkinder reviewed (approved):
no doc changes to review, Kconfig changes look OK.

MarekPieta commented:
@pdunaj, could you review ?

pdunaj commented on subsys/debug/tracing/cpu_stats.c:
```diff
@@ -0,0 +1,206 @@
+	CPU_STATE_SCHEDULER
+};
+
+static volatile enum cpu_state last_cpu_state = CPU_STATE_SCHEDULER;
```
misused volatile

pdunaj replied:
also in other places

MarekPieta replied:
removed

pdunaj commented on subsys/debug/tracing/cpu_stats.c:
```diff
@@ -0,0 +1,206 @@
+	switch (last_cpu_state) {
+	case CPU_STATE_IDLE:
+		increase_counter(&idle_cnt);
+	break;
```
please indent break

MarekPieta replied:
done

pdunaj commented on subsys/debug/tracing/cpu_stats.c:
```diff
@@ -0,0 +1,206 @@
+
+void sys_trace_isr_enter(void)
+{
+	if (nested_interrupts == 0) {
```
this is very tricky note that you can be preempted by another interrupt before you increment the `nested_interrupts` (unless interrupts were disabled but I don't think they are). You cannot simply use atomic increment or cmpxch as you may end up with counters updated after the high priority interrupts are done. I think the only option is to lock interrupts for entire operation. (Note that you already lock interrupts while updating the counters.)

MarekPieta replied:
done

pdunaj commented on subsys/debug/tracing/cpu_stats.c:
```diff
@@ -0,0 +1,206 @@
+
+void sys_trace_isr_exit(void)
+{
+	nested_interrupts--;
```
check comment above - lock irq seems needed

MarekPieta replied:
done

MarekPieta commented:
I made requested changes

pdunaj commented on subsys/debug/tracing/cpu_stats.c:
```diff
@@ -0,0 +1,209 @@
+
+void sys_trace_thread_switched_in(void)
+{
+	__ASSERT_NO_MSG(nested_interrupts == 0);
```
I would lock it too. Otherwise interrupt can come and harm us. (Even if upper layer disables interrupts, if we depend on them being disabled we should either disable them or document the requirement and assert state). I prefer adding a lock - it won't increase the cost of call much as lock is already taken when counters are updated.

MarekPieta replied:
done

pdunaj commented:
hi @mped-oticon , can you review so we could start using the change in the project? thanks!

mped-oticon commented:
lgtm

pdunaj commented:
hi @carlescufi , could you merge this one?

nashif commented on subsys/debug/Kconfig:
```diff
@@ -189,4 +189,19 @@ config OPENOCD_SUPPORT
+	select TRACING
+
+config TRACING_CPU_STATS_LOG
+	bool "Logging current CPU usage"
```
"Enable logging for current CPU usage"

MarekPieta replied:
done

nashif commented on subsys/debug/Kconfig:
```diff
@@ -189,4 +189,19 @@ config OPENOCD_SUPPORT
+
+config TRACING_CPU_STATS_LOG
+	bool "Logging current CPU usage"
+	depends on TRACING_CPU_STATS
```
add help message

MarekPieta replied:
done

nashif commented on subsys/debug/Kconfig:
```diff
@@ -189,4 +189,19 @@ config OPENOCD_SUPPORT
+	depends on TRACING_CPU_STATS
+
+config TRACING_CPU_STATS_INTERVAL
+	int "Logging interval for CPU measurements [ms]"
```
"Enable ...."

MarekPieta replied:
done

nashif replied:
ok, it is fine as is.

nashif commented on subsys/debug/Kconfig:
```diff
@@ -189,4 +189,19 @@ config OPENOCD_SUPPORT
+config TRACING_CPU_STATS_INTERVAL
+	int "Logging interval for CPU measurements [ms]"
+	default 2000
+	depends on TRACING_CPU_STATS_LOG
```
add a help message please

MarekPieta replied:
done

nashif commented on subsys/debug/tracing/include/tracing_cpu_stats.h:
```diff
@@ -0,0 +1,40 @@
+#include <kernel.h>
+#include <kernel_structs.h>
+#include <init.h>
+#include <misc/printk.h>
```
why do you need printk here?

MarekPieta replied:
Currently printk is used to periodically display information about CPU usage.

nashif replied:
it should be included where it is being used.

MarekPieta replied:
done

MarekPieta commented on subsys/debug/Kconfig:
```diff
@@ -189,4 +189,19 @@ config OPENOCD_SUPPORT
 	  OpenOCD to determine the state of running threads.  (This option
 	  selects CONFIG_THREAD_MONITOR, so all of its caveats are implied.)
 
+config TRACING_CPU_STATS
```
done

carlescufi commented:
@nashif can you take another look?

nashif commented:
> @nashif can you take another look?

did, wonder if the printk.h inclusion would be moved to where printk is being used.

MarekPieta commented:
@nashif I made requested change

pizi-nordic commented on subsys/debug/tracing/cpu_stats.c:
```diff
@@ -0,0 +1,214 @@
+extern k_tid_t const _idle_thread;
+#endif
+
+static inline int is_idle_thread(struct k_thread *thread)
```
There is already `_is_idle()` in sched.c.
IMHO we should re-use the code instead of copying it around.

MarekPieta replied:
this function is static inline, so we cannot simply use it in our files;
it is done like this for sysview profiler hooks

pdunaj replied:
if we can limit the number of calls at this level it would be great, so I recommend to leave it as is.

pizi-nordic replied:
Could we move the `_is_idle()` to some header and make it public for other kernel parts?
@andrewboie: What do you think?

nashif replied:
i already asked @andyross to make this a function call. We can clean this up once this function is available.

pdunaj replied:
@pizi-nordic let's fix it later when API is changed. We need this change.

pizi-nordic commented on subsys/debug/tracing/cpu_stats.c:
```diff
@@ -0,0 +1,214 @@
+#endif
+}
+
+void increase_counter(volatile u64_t *cnt)
```
What about update_counter()? Then the naming will match the cpu_stats_update_counters().

MarekPieta replied:
done

pizi-nordic commented on subsys/debug/tracing/cpu_stats.c:
```diff
@@ -0,0 +1,214 @@
+		break;
+
+	default:
+		__ASSERT_NO_MSG(false);
```
Please add comment what is handled here (invalid cpu state?).

MarekPieta replied:
done

pizi-nordic commented on subsys/debug/tracing/cpu_stats.c:
```diff
@@ -0,0 +1,214 @@
+	}
+}
+
+u64_t cpu_stats_sched_get_ns(void)
```
Could we replace these function by `cpu_stats_get(struct cpu_stats *stats)`?
IMHO such approach will be simpler and easier to extend.

MarekPieta replied:
done
